src/legacy/shrink.rs

```diff
@@ -0,0 +1,379 @@
+use std::collections::VecDeque;
+use std::io::{self, copy, Error, Read};
+
+use bitstream_io::{BitRead, BitReader, Endianness, LittleEndian};
+
+const MIN_CODE_SIZE: u8 = 9;
+const MAX_CODE_SIZE: u8 = 13;
+
+const MAX_CODE: usize = (1 << MAX_CODE_SIZE) - 1;
+const CONTROL_CODE: usize = 256;
+const INC_CODE_SIZE: u16 = 1;
+const PARTIAL_CLEAR: u16 = 2;
+
+// const HASH_BITS: usize = MAX_CODE_SIZE + 1; /* For a load factor of 0.5. */
+// const HASHTAB_SIZE: usize = 1 << HASH_BITS;
+const UNKNOWN_LEN: u16 = u16::MAX;
+
+struct CodeQueue {
+    next_idx: usize,
+    codes: [Option<u16>; MAX_CODE as usize - CONTROL_CODE + 1],
+}
+
+impl CodeQueue {
+    fn new() -> Self {
+        let mut codes = [None; MAX_CODE as usize - CONTROL_CODE + 1];
+        for (i, code) in (CONTROL_CODE as u16 + 1..=MAX_CODE as u16).enumerate() {
+            codes[i] = Some(code);
+        }
+        Self { next_idx: 0, codes }
+    }
+
+    // Return the next code in the queue, or INVALID_CODE if the queue is empty.
+    fn next(&self) -> Option<u16> {
+        //   assert(q->next_idx < sizeof(q->codes) / sizeof(q->codes[0]));
+        self.codes[self.next_idx]
+    }
+
+    /// Return and remove the next code from the queue, or return INVALID_CODE if
+    /// the queue is empty.
+    fn remove_next(&mut self) -> Option<u16> {
+        let res = self.next();
+        if res.is_some() {
+            self.next_idx += 1;
+        }
+        res
+    }
+}
+
+#[derive(Clone, Debug, Copy)]
+struct Codetab {
+    prefix_code: Option<u16>,
+    ext_byte: u8,
+    len: u16,
+    last_dst_pos: usize,
+}
+
+impl Default for Codetab {
+    fn default() -> Self {
+        Self {
+            prefix_code: None,
+            ext_byte: 0,
+            len: 0,
+            last_dst_pos: 0,
+        }
+    }
+}
+
+impl Codetab {
+    pub fn create_new() -> [Self; MAX_CODE + 1] {
```

Comment:
By convention this should be called `new_array()` since it doesn't simply return Self, and `new()` if it did.